File: cocos2d/core/utils/trans-pool/mem-pool.js
/****************************************************************************
 Copyright (c) 2019 Xiamen Yaji Software Co., Ltd.

 https://www.cocos.com/

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated engine source code (the "Software"), a limited,
 worldwide, royalty-free, non-assignable, revocable and non-exclusive license
 to use Cocos Creator solely to develop games on your target platforms. You shall
 not use Cocos Creator software for developing other software or tools that's
 used for developing games. You are not granted to publish, distribute,
 sublicense, and/or sell copies of Cocos Creator.

 The software or tools in this License Agreement are licensed, not sold.
 Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/

let MemPool = function (unitClass) {
    this._unitClass = unitClass;
    this._pool = [];
    this._findOrder = [];

    if (CC_JSB && CC_NATIVERENDERER) {
        this._initNative();
    }
};

let proto = MemPool.prototype;
proto._initNative = function () {
    this._nativeMemPool = new renderer.MemPool();
};

proto._buildUnit = function (unitID) {
    let unit = new this._unitClass(unitID, this);
    if (CC_JSB && CC_NATIVERENDERER) {
        this._nativeMemPool.updateCommonData(unitID, unit._data, unit._signData);
    }
    return unit;
};

proto._destroyUnit = function (unitID) {
    this._pool[unitID] = null;
    for (let idx = 0, n = this._findOrder.length; idx < n; idx++) {
        let unit = this._findOrder[idx];
        if (unit && unit.unitID == unitID) {
            this._findOrder.splice(idx, 1);
            break;
        }
    }
    if (CC_JSB && CC_NATIVERENDERER) {
        this._nativeMemPool.removeCommonData(unitID);
    }
};

proto._findUnitID = function () {
    let unitID = 0;
    let pool = this._pool;
    while (pool[unitID]) unitID++;
    return unitID;
};

proto.pop = function () {
    let findUnit = null;
    let idx = 0;
    let findOrder = this._findOrder;
    let pool = this._pool;
    for (let n = findOrder.length; idx < n; idx++) {
        let unit = findOrder[idx];
        if (unit && unit.hasSpace()) {
            findUnit = unit;
            break;
        }
    }

    if (!findUnit) {
        let unitID = this._findUnitID();
        findUnit = this._buildUnit(unitID);
        pool[unitID] = findUnit;
        findOrder.push(findUnit);
        idx = findOrder.length - 1;
    }

    // swap has space unit to first position, so next find will fast
    let firstUnit = findOrder[0];
    if (firstUnit !== findUnit) {
        findOrder[0] = findUnit;
        findOrder[idx] = firstUnit;
    }

    return findUnit.pop();
};

proto.push = function (info) {
    let unit = this._pool[info.unitID];
    unit.push(info.index);
    if (this._findOrder.length > 1 && unit.isAllFree()) {
        this._destroyUnit(info.unitID);
    }
    return unit;
};
module.exports = MemPool;

File: native/cocos/renderer/scene/MemPool.cpp
/****************************************************************************
 Copyright (c) 2018 Xiamen Yaji Software Co., Ltd.
 
 http://www.cocos2d-x.org
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/

#include "MemPool.hpp"
#include "cocos/scripting/js-bindings/jswrapper/SeApi.h"
#include "base/ccMacros.h"

RENDERER_BEGIN

UnitBase::UnitBase()
{
    
}

UnitBase::~UnitBase()
{
    
}

void UnitBase::set(se::Object** dataObj, uint8_t** data, std::size_t* dataLen, se::Object* jsData)
{
    if (*dataObj == jsData) return;
    
    if (*dataObj)
    {
        (*dataObj)->unroot();
        (*dataObj)->decRef();
        *dataObj = nullptr;
    }
    
    if (jsData == nullptr) return;
    
    *dataObj = jsData;
    (*dataObj)->root();
    (*dataObj)->incRef();
    *data = nullptr;
    *dataLen = 0;
    (*dataObj)->getTypedArrayData(data, dataLen);
}

void UnitBase::unset(se::Object** dataObj, uint8_t** data, std::size_t* dataLen)
{
    if (*dataObj)
    {
        (*dataObj)->unroot();
        (*dataObj)->decRef();
        *dataObj = nullptr;
    }
    
    *data = nullptr;
    *dataLen = 0;
}

UnitCommon::UnitCommon()
{
    
}

UnitCommon::~UnitCommon()
{
    unset(&dataObj, (uint8_t**)&data, &dataLen);
    unset(&signDataObj, (uint8_t**)&signData, &signDataLen);
}

void UnitCommon::setData(se::Object* jsData)
{
    set(&dataObj, (uint8_t**)&data, &dataLen, jsData);
}

void UnitCommon::setSignData(se::Object* jsSignData)
{
    set(&signDataObj, (uint8_t**)&signData, &signDataLen, jsSignData);
}

MemPool::MemPool()
{
    
}

MemPool::~MemPool()
{
    for(auto it = _commonPool.begin(); it != _commonPool.end(); it++)
    {
        if (*it)
        {
            delete (*it);
        }
    }
    _commonPool.clear();
}

void MemPool::removeCommonData(std::size_t unitID)
{
    CCASSERT(unitID < _commonPool.size(), "MemPool removeCommonData unitID can not be rather than pool size");
    auto unit = _commonPool[unitID];
    if (unit) 
    {
        for (auto it = _commonList.begin(); it != _commonList.end(); it++)
        {
            if ((*it)->unitID == unitID)
            {
                _commonList.erase(it);
                break;
            }
        }
        delete unit;
        _commonPool[unitID] = nullptr;
    }
}

void MemPool::updateCommonData(std::size_t unitID, se_object_ptr dataObj, se_object_ptr signDataObj)
{
    CCASSERT(unitID <= _commonPool.size(), "MemPool updateData unitID can not be rather than pool size");
    
    UnitCommon* unit = nullptr;
    if (unitID == _commonPool.size())
    {
        unit = new UnitCommon;
        _commonPool.push_back(unit);
        _commonList.push_back(unit);
    }
    else if (unitID < _commonPool.size())
    {
        unit = _commonPool[unitID];
        if (!unit) 
        {
            unit = new UnitCommon;
            _commonPool[unitID] = unit;
            _commonList.push_back(unit);
        }
    }
    else
    {
        return;
    }
    
    unit->unitID = unitID;
    unit->setData(dataObj);
    unit->setSignData(signDataObj);
}

const std::vector<UnitCommon*>& MemPool::getCommonPool() const
{
    return _commonPool;
}

const std::vector<UnitCommon*>& MemPool::getCommonList() const
{
    return _commonList;
}

UnitCommon* MemPool::getCommonUnit(std::size_t unitID)
{
    CCASSERT(unitID < _commonPool.size(), "MemPool getCommonUnit unitID can not be rather than pool size");
    return _commonPool[unitID];
}

RENDERER_END
